### PR TITLE
Changes needed for consensus integration

### PR DESCRIPTION
**Pull request by nc6**
This PR incorporates the various changes needed for integration with the consensus layer. They can be divided into three basic ideas:

- Firstly, we expose a bunch of things which were previously hidden, because they're needed.
- Secondly, we change some of the serialisation code to take a more explicit serialiser, rather than using `ToCBOR` constraints which the consensus layer can typically not satisfy.
- Finally, we split some of the rule processing to fit with an explicit header/body split.

### CODE REVIEW

File: crypto/test/Test/Cardano/Crypto/Signing/Signing.hs [hunk 1]
```diff
@@ -35,8 +36,8 @@ prop_sign = property $ do
   a        <- forAll genData
 
   assert
-    $ verifySignature Dummy.protocolMagicId SignForTestingOnly pk a
-    $ sign Dummy.protocolMagicId SignForTestingOnly sk a
+    $ verifySignature toCBOR Dummy.dummyProtocolMagicId SignForTestingOnly pk a
```
Comment (ruhatch) [inline crypto/test/Test/Cardano/Crypto/Signing/Signing.hs:39]: Should be just `protocolMagicId` now

File: crypto/test/Test/Cardano/Crypto/Signing/Signing.hs [hunk 2]
```diff
@@ -35,8 +36,8 @@ prop_sign = property $ do
...
+    $ sign dummyProtocolMagicId SignForTestingOnly sk a
```
Comment (ruhatch) [inline crypto/test/Test/Cardano/Crypto/Signing/Signing.hs:40]: This should be `Dummy.protocolMagicId`

**Review by ruhatch — APPROVED** [review]
LGTM! Is the `HeaderState` stuff still good given the header/body split issues?

**Comment by nc6** [comment]
@ruhatch To answer your comment, no! I'm probably going to drop the last commit. So this PR isn't ready for merge yet...

**Comment by nc6** [comment]
Closing in favour of #500

**Pull request closed without merging**